src/query/functions/unconsolidated/timestamp.go

```diff
@@ -108,27 +111,27 @@ func (n *timestampNode) ProcessBlock(queryCtx *models.QueryContext, ID parser.No
 		return nil, err
 	}
 
-	if err = builder.AddCols(iter.StepCount()); err != nil {
+	count := iter.StepCount()
+	seriesCount := len(iter.SeriesMeta())
+	if err = builder.AddCols(count); err != nil {
 		return nil, err
 	}
 
+	bounds := b.Meta().Bounds
+	currentStep := float64(bounds.Start.Unix())
+	step := float64(bounds.StepSize) / float64(time.Second)
+	values := make([]float64, seriesCount)
 	for index := 0; iter.Next(); index++ {
-		step := iter.Current()
-		values := make([]float64, len(step.Values()))
-		ts.Memset(values, math.NaN())
-		for i, dps := range step.Values() {
-			if len(dps) == 0 {
-				continue
+		curr := iter.Current()
+		util.Memset(values, currentStep)
+		for i, dp := range curr.Values() {
+			if math.IsNaN(dp) {
+				values[i] = math.NaN()
```

Comment:
Would you be open to putting a comment here why if it's NaN here this series should be NaN instead of the timestamp value?